modules/python/python-http-header.c

```diff
@@ -0,0 +1,383 @@
+                evt_tag_str("class", self->class),
+                evt_tag_str("method", "get_headers"));
+    }
+
+exit:
+  Py_XDECREF(py_args);
+  return self->py.get_headers != NULL;
+}
+
+static void
+_py_detach_bindings(PythonHttpHeaderPlugin *self)
+{
+  Py_CLEAR(self->py.class);
+  Py_CLEAR(self->py.instance);
+  Py_CLEAR(self->py.get_headers);
+}
+
+static void
+_append_str_to_list(gpointer data, gpointer user_data)
+{
+  List *list = (List *) user_data;
+  list_append(list, data);
+}
+
+static void
+_append_headers(PythonHttpHeaderPlugin *self, HttpHeaderRequestSignalData *data)
+{
+  PyObject *py_ret = NULL,
+            *py_list = NULL,
+             *py_args = NULL,
+              *py_ret_list = NULL;
+
+  time_t now = time(NULL);
+  if ((now - self->last_run) < self->timeout)
+    goto exit;
+
+  // The GIL also guards non-Python plugin struct members from concurrent changes below.
+  PyGILState_STATE gstate = PyGILState_Ensure();
+
+  self->timeout = 0;
+  g_list_free_full(self->last_headers, g_free);
+  self->last_headers = NULL;
+
+  py_list = _py_convert_list_to_pylist(data->request_headers);
+  if (!py_list)
+    goto cleanup;
+
+  py_args = Py_BuildValue("(sO)", data->request_body->str, py_list);
+  if (!py_args)
+    {
+      gchar buf[256];
+
+      msg_error("Error creating Python arguments",
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto cleanup;
+    }
+
+  py_ret = _py_invoke_function_with_args(self->py.get_headers, py_args, self->class, "_append_headers");
+  if (!py_ret || !PyArg_ParseTuple(py_ret, "iO", &self->timeout, &py_ret_list))
+    {
+      gchar buf[256];
+
+      msg_error("Invalid response returned by Python call",
+                evt_tag_str("class", self->class),
+                evt_tag_str("method", "get_headers"),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto cleanup;
+    }
+
+  msg_debug("Python call returned valid response",
+            evt_tag_str("class", self->class),
+            evt_tag_str("method", "get_headers"),
+            evt_tag_str("return_type", "Tuple"),
+            evt_tag_int("len", PyTuple_Size(py_ret)));
+
+  if (!_py_append_pylist_to_list(py_ret_list, &self->last_headers))
+    {
+      gchar buf[256];
+
+      msg_error("Converting Python List failed",
+                evt_tag_str("class", self->class),
+                evt_tag_str("method", "get_headers"),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+    }
+
+cleanup:
+  Py_XDECREF(py_args);
+  Py_XDECREF(py_ret);
+  Py_XDECREF(py_list);
+  Py_XDECREF(py_ret_list);
+
+  self->last_run = now;
+
+  PyGILState_Release(gstate);
+
+exit:
+  g_assert(self->last_headers);
```

Comment:
This cannot be an exception. In case of error, we can get here with `self->last_headers == NULL`.